src/xGadgetron/cGadgetron/include/sirf/Gadgetron/TrajectoryPreparation.h

```diff
@@ -49,26 +49,108 @@ namespace sirf{
+    {
+        if(mr_acq.number() <= 0){
+            throw std::runtime_error("Please pass a non-empty container.");
+        }
+        
+        ISMRMRD::Acquisition acq;
+        mr_acq.get_acquisition(0, acq);
+
+        if( acq.trajectory_dimensions() != D){
+            throw std::runtime_error("Please give Acquisition with a the correct dimensionality if you want to use it here.");
+        }
+        
+        std::vector<size_t> kspace_dims;
+        mr_acq.get_kspace_dimensions(kspace_dims);
+
+        size_t const Nz = mr_acq.acquisitions_info().get_IsmrmrdHeader().encoding[0].encodedSpace.matrixSize.z;
+        size_t const num_traj_points = unique_kz ? mr_acq.number()*kspace_dims[0]/Nz 
+                                                 : mr_acq.number()*kspace_dims[0];
+
+        TrajPointSet traj;
+        
+        if(unique_kz)
+            traj.resize(num_traj_points);
+            
+        for(int ia=0; ia<mr_acq.number(); ++ia)
+        {
+            mr_acq.get_acquisition(ia, acq);
+            
+            if(unique_kz)
+            {
+                if(acq.idx().kspace_encode_step_2 != 0)
+                    continue;
+                    
+                get_ky_sorted_trajectory(traj, acq);
+            }
+            else
+                append_to_trajectory(traj, acq);
+        }
+
+        return traj;
+    }
 
-    void update_acquisitions_info(sirf::MRAcquisitionData& mr_acq);
 
+protected:
     ISMRMRD::Encoding kspace_encoding_;
     ISMRMRD::TrajectoryType traj_type_;
 
-    uint16_t traj_dim_;
-    virtual void set_acquisition_trajectory(ISMRMRD::Acquisition& acq) const =0;
-    virtual std::vector<float> calculate_trajectory(ISMRMRD::Acquisition& acq) const =0;
+    virtual void set_acquisition_trajectory(ISMRMRD::Acquisition& acq) const
+    {
+        acq.resize(acq.number_of_samples(),acq.active_channels(), D);
+        TrajPointSet acq_traj = this->calculate_trajectory(acq);
+        acq.setTraj(&(acq_traj[0][0]));
+    }
+    virtual void update_acquisitions_info(sirf::MRAcquisitionData& mr_acq)
+    {
+        ISMRMRD::IsmrmrdHeader hdr = mr_acq.acquisitions_info().get_IsmrmrdHeader();
+
+        if(hdr.encoding.size() != 1)
+            throw LocalisedException("Currrently only files with one encoding are supported", __FILE__, __LINE__);
+
+        hdr.encoding[0].trajectory = traj_type_;
+        kspace_encoding_ = hdr.encoding[0];
+
+        std::stringstream hdr_stream;
+        serialize(hdr, hdr_stream);
+
+        AcquisitionsInfo ai(hdr_stream.str());
+        mr_acq.set_acquisitions_info(ai);
+    }
+
+    virtual TrajPointSet calculate_trajectory(ISMRMRD::Acquisition& acq) const =0;
+    virtual void append_to_trajectory(TrajPointSet& tps, ISMRMRD::Acquisition& acq)=0;
+    virtual void get_ky_sorted_trajectory(TrajPointSet& tps, ISMRMRD::Acquisition& acq)=0;
+
 };
 
+typedef TrajectoryPreparation<2> TrajPrep2D;
+typedef TrajectoryPreparation<3> TrajPrep3D;
```

Comment:
I'd rather call them `TrajectoryPreparation2D` etc. It otherwise becomes hard to remember when you used the abbreviation and when not.